File: backend/src/GatherGrove.Application/Services/MembershipTypeService.cs
using Microsoft.EntityFrameworkCore;
using GatherGrove.Application.DTOs;
using GatherGrove.Domain.Entities;
using GatherGrove.Infrastructure.Data;

namespace GatherGrove.Application.Services;

/// <summary>
/// Service for membership type operations
/// </summary>
public class MembershipTypeService : IMembershipTypeService
{
    private readonly GatherGroveDbContext _context;

    public MembershipTypeService(GatherGroveDbContext context)
    {
        _context = context;
    }

    /// <summary>
    /// Creates a new membership type for a club
    /// </summary>
    public async Task<MembershipTypeResponse> CreateMembershipTypeAsync(int clubId, CreateMembershipTypeRequest request)
    {
        // Check if club exists
        var club = await _context.Clubs.FindAsync(clubId);
        if (club == null)
        {
            throw new ArgumentException($"Club with ID {clubId} not found", nameof(clubId));
        }

        // Check if membership type name already exists in this club
        var existingType = await _context.MembershipTypes
            .FirstOrDefaultAsync(mt => mt.ClubId == clubId && mt.Name == request.Name);

        if (existingType != null)
        {
            throw new ArgumentException($"A membership type with the name '{request.Name}' already exists in this club");
        }

        var now = DateTime.UtcNow;
        var membershipType = new MembershipType
        {
            ClubId = clubId,
            Name = request.Name,
            Description = request.Description,
            DuesAmount = request.DuesAmount,
            DuesFrequency = request.DuesFrequency,
            IsActive = true,
            CreatedAt = now,
            UpdatedAt = now
        };

        _context.MembershipTypes.Add(membershipType);
        await _context.SaveChangesAsync();

        return new MembershipTypeResponse
        {
            Id = membershipType.Id,
            ClubId = membershipType.ClubId,
            Name = membershipType.Name,
            Description = membershipType.Description,
            DuesAmount = membershipType.DuesAmount,
            DuesFrequency = membershipType.DuesFrequency,
            IsActive = membershipType.IsActive,
            CreatedAt = membershipType.CreatedAt,
            UpdatedAt = membershipType.UpdatedAt,
            MemberCount = 0 // New membership type has no members
        };
    }

    /// <summary>
    /// Gets all membership types for a club
    /// </summary>
    public async Task<List<MembershipTypeResponse>> GetMembershipTypesByClubAsync(int clubId)
    {
        var membershipTypes = await _context.MembershipTypes
            .Where(mt => mt.ClubId == clubId)
            .OrderBy(mt => mt.Name)
            .AsNoTracking()
            .ToListAsync();

        // Get member counts for each membership type
        var membershipTypeIds = membershipTypes.Select(mt => mt.Id).ToList();
        var memberCounts = await _context.Members
            .Where(m => membershipTypeIds.Contains(m.MembershipTypeId) && m.Status == "Active")
            .GroupBy(m => m.MembershipTypeId)
            .Select(g => new { MembershipTypeId = g.Key, Count = g.Count() })
            .AsNoTracking()
            .ToListAsync();

        var memberCountDict = memberCounts.ToDictionary(mc => mc.MembershipTypeId, mc => mc.Count);

        return membershipTypes.Select(mt => new MembershipTypeResponse
        {
            Id = mt.Id,
            ClubId = mt.ClubId,
            Name = mt.Name,
            Description = mt.Description,
            DuesAmount = mt.DuesAmount,
            DuesFrequency = mt.DuesFrequency,
            IsActive = mt.IsActive,
            CreatedAt = mt.CreatedAt,
            UpdatedAt = mt.UpdatedAt,
            MemberCount = memberCountDict.GetValueOrDefault(mt.Id, 0)
        }).ToList();
    }

    /// <summary>
    /// Gets a specific membership type by ID
    /// </summary>
    public async Task<MembershipTypeResponse?> GetMembershipTypeByIdAsync(int clubId, int membershipTypeId)
    {
        var membershipType = await _context.MembershipTypes
            .Where(mt => mt.ClubId == clubId && mt.Id == membershipTypeId)
            .AsNoTracking()
            .FirstOrDefaultAsync();

        if (membershipType == null)
        {
            return null;
        }

        // Get member count for this membership type
        var memberCount = await _context.Members
            .CountAsync(m => m.MembershipTypeId == membershipType.Id && m.Status == "Active");

        return new MembershipTypeResponse
        {
            Id = membershipType.Id,
            ClubId = membershipType.ClubId,
            Name = membershipType.Name,
            Description = membershipType.Description,
            DuesAmount = membershipType.DuesAmount,
            DuesFrequency = membershipType.DuesFrequency,
            IsActive = membershipType.IsActive,
            CreatedAt = membershipType.CreatedAt,
            UpdatedAt = membershipType.UpdatedAt,
            MemberCount = memberCount
        };
    }

    /// <summary>
    /// Updates an existing membership type
    /// </summary>
    public async Task<MembershipTypeResponse> UpdateMembershipTypeAsync(int clubId, int membershipTypeId, UpdateMembershipTypeRequest request)
    {
        var membershipType = await _context.MembershipTypes
            .FirstOrDefaultAsync(mt => mt.ClubId == clubId && mt.Id == membershipTypeId);

        if (membershipType == null)
        {
            throw new ArgumentException($"Membership type with ID {membershipTypeId} not found in club {clubId}");
        }

        // Check if the new name conflicts with an existing membership type (excluding current one)
        var existingType = await _context.MembershipTypes
            .FirstOrDefaultAsync(mt => mt.ClubId == clubId && mt.Name == request.Name && mt.Id != membershipTypeId);

        if (existingType != null)
        {
            throw new ArgumentException($"A membership type with the name '{request.Name}' already exists in this club");
        }

        // Update properties
        membershipType.Name = request.Name;

        // Only update Description if provided (not empty)
        if (!string.IsNullOrEmpty(request.Description))
        {
            membershipType.Description = request.Description;
        }

        membershipType.DuesAmount = request.DuesAmount;

        // Only update DuesFrequency if provided (not empty)
        if (!string.IsNullOrEmpty(request.DuesFrequency))
        {
            membershipType.DuesFrequency = request.DuesFrequency;
        }

        membershipType.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();

        // Get member count for this membership type
        var memberCount = await _context.Members
            .CountAsync(m => m.MembershipTypeId == membershipType.Id && m.Status == "Active");

        return new MembershipTypeResponse
        {
            Id = membershipType.Id,
            ClubId = membershipType.ClubId,
            Name = membershipType.Name,
            Description = membershipType.Description,
            DuesAmount = membershipType.DuesAmount,
            DuesFrequency = membershipType.DuesFrequency,
            IsActive = membershipType.IsActive,
            CreatedAt = membershipType.CreatedAt,
            UpdatedAt = membershipType.UpdatedAt,
            MemberCount = memberCount
        };
    }

    /// <summary>
    /// Deletes a membership type
    /// </summary>
    public async Task<bool> DeleteMembershipTypeAsync(int clubId, int membershipTypeId)
    {
        var membershipType = await _context.MembershipTypes
            .FirstOrDefaultAsync(mt => mt.ClubId == clubId && mt.Id == membershipTypeId);

        if (membershipType == null)
        {
            return false;
        }

        // Check if any members are assigned to this membership type
        var memberCount = await _context.Members
            .CountAsync(m => m.MembershipTypeId == membershipTypeId);

        if (memberCount > 0)
        {
            var memberText = memberCount == 1 ? "member" : "members";
            throw new InvalidOperationException($"Cannot delete membership type because it is assigned to {memberCount} {memberText}. Please reassign these members to a different membership type before deletion.");
        }

        _context.MembershipTypes.Remove(membershipType);
        await _context.SaveChangesAsync();

        return true;
    }
}

File: client/src/services/membershipTypeService.ts
import apiClient from './apiClient';
import { ErrorHandler } from '@/lib/errorHandler';
import { DuesFrequency } from '@/constants/duesFrequency';

// Types for membership types
export interface CreateMembershipTypeRequest {
  name: string;
  description?: string;
  duesAmount: number;
  duesFrequency: DuesFrequency;
}

export interface UpdateMembershipTypeRequest {
  name: string;
  description?: string;
  duesAmount: number;
  duesFrequency?: DuesFrequency;
}

export interface MembershipTypeResponse {
  id: number;
  clubId: number;
  name: string;
  description: string;
  duesAmount: number;
  duesFrequency: DuesFrequency;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  memberCount: number;
}

/**
 * Service for managing membership types
 */
class MembershipTypeService {
  /**
   * Creates a new membership type for a club
   * @param clubId - The ID of the club
   * @param membershipType - Membership type details
   * @returns Promise with created membership type
   */
  async createMembershipType(clubId: number, membershipType: CreateMembershipTypeRequest): Promise<MembershipTypeResponse> {
    try {
      const response = await apiClient.post<MembershipTypeResponse>(
        `/clubs/${clubId}/membership-types`, 
        membershipType
      );
      return response.data;
    } catch (error) {
      throw ErrorHandler.handleApiError(error, {
        context: 'creating membership type',
        action: 'Please check the membership type details and try again',
        customMessages: {
          400: 'Please verify all membership type fields are correct',
          403: 'You do not have permission to create membership types',
          409: `A membership type named "${membershipType.name}" already exists`
        }
      });
    }
  }

  /**
   * Gets all membership types for a club
   * @param clubId - The ID of the club
   * @returns Promise with list of membership types
   */
  async getMembershipTypes(clubId: number): Promise<MembershipTypeResponse[]> {
    try {
      const response = await apiClient.get<MembershipTypeResponse[]>(
        `/clubs/${clubId}/membership-types`
      );
      
      // The API returns a direct array
      const data = response.data;
      
      // Ensure we always return an array
      return Array.isArray(data) ? data : [];
    } catch (error) {
      throw ErrorHandler.handleApiError(error, {
        context: 'loading membership types',
        action: 'Please try refreshing the page',
        customMessages: {
          403: 'You do not have permission to view membership types',
          404: 'Club not found or you do not have access to it'
        }
      });
    }
  }

  /**
   * Gets a specific membership type by ID
   * @param clubId - The ID of the club
   * @param membershipTypeId - The ID of the membership type
   * @returns Promise with membership type details
   */
  async getMembershipType(clubId: number, membershipTypeId: number): Promise<MembershipTypeResponse> {
    try {
      const response = await apiClient.get<MembershipTypeResponse>(
        `/clubs/${clubId}/membership-types/${membershipTypeId}`
      );
      return response.data;
    } catch (error) {
      throw ErrorHandler.handleApiError(error, {
        context: 'loading membership type details',
        action: 'Please try refreshing the page or go back to the membership types list',
        customMessages: {
          404: 'Membership type not found or has been deleted',
          403: 'You do not have permission to view this membership type'
        }
      });
    }
  }

  /**
   * Updates an existing membership type
   * @param clubId - The ID of the club
   * @param membershipTypeId - The ID of the membership type to update
   * @param membershipType - Updated membership type details
   * @returns Promise with updated membership type
   */
  async updateMembershipType(clubId: number, membershipTypeId: number, membershipType: UpdateMembershipTypeRequest): Promise<MembershipTypeResponse> {
    try {
      const response = await apiClient.put<MembershipTypeResponse>(
        `/clubs/${clubId}/membership-types/${membershipTypeId}`, 
        membershipType
      );
      return response.data;
    } catch (error) {
      throw ErrorHandler.handleApiError(error, {
        context: 'updating membership type',
        action: 'Please check the membership type details and try again',
        customMessages: {
          400: 'Please verify all membership type fields are correct',
          404: 'Membership type not found or has been deleted',
          403: 'You do not have permission to edit this membership type',
          409: `A membership type named "${membershipType.name}" already exists`
        }
      });
    }
  }

  /**
   * Deletes a membership type
   * @param clubId - The ID of the club
   * @param membershipTypeId - The ID of the membership type to delete
   * @returns Promise that resolves when deletion is complete
   */
  async deleteMembershipType(clubId: number, membershipTypeId: number): Promise<void> {
    try {
      await apiClient.delete(`/clubs/${clubId}/membership-types/${membershipTypeId}`);
    } catch (error) {
      throw ErrorHandler.handleApiError(error, {
        context: 'deleting membership type',
        action: 'Please reassign members to a different membership type before attempting deletion',
        customMessages: {
          404: 'Membership type not found or has already been deleted',
          403: 'You do not have permission to delete this membership type'
        }
      });
    }
  }
}

// Export singleton instance
const membershipTypeService = new MembershipTypeService();
export default membershipTypeService;
export { membershipTypeService }; 